compiler/p/codegen/OMRMemoryReference.cpp

```diff
@@ -1068,375 +1068,272 @@ void OMR::Power::MemoryReference::mapOpCode(TR::Instruction *currentInstruction)
-            // we don't need to use the index reg
-            // just use 0 as RA, base as RB
-            base->setRegisterFieldRB(wcursor);
+            // To implicitly use a displacement of 0, put the base register into the index position and use r0 as the
+            // base register.
+            self()->setIndexRegister(base);
+            self()->setBaseRegister(cg->machine()->getRealRegister(TR::RealRegister::gr0));
             }
          }
       else
          {
-         base->setRegisterFieldRA(wcursor);
-         index->setRegisterFieldRB(wcursor);
-         }
-      }
-   else
-      {
-      int32_t  displacement = self()->getOffset(*comp);
-      uint32_t preserve = *wcursor;
-      int32_t  lower = LO_VALUE(displacement) & 0x0000ffff;
-      int32_t  upper = cg->hiValue(displacement) & 0x0000ffff;
+         TR_ASSERT_FATAL_WITH_INSTRUCTION(
+            currentInstruction,
+            self()->isBaseModifiable(),
+            "Delayed index form without index register requires isBaseModifiable"
+         );
 
-      if (self()->isUsingDelayedIndexedForm())
-         {
-         // must be isBaseModifiable()
-         if (displacement<LOWER_IMMED || displacement>UPPER_IMMED)
-            {
-            *wcursor = 0x3c000000 | upper;
-            base->setRegisterFieldRT(wcursor);
-            base->setRegisterFieldRA(wcursor);
-            cursor += PPC_INSTRUCTION_LENGTH;
-            wcursor++;                                  // addis Rb,Rb,upper
-
-            *wcursor = 0x38000000 | lower;
-            base->setRegisterFieldRT(wcursor);
-            base->setRegisterFieldRA(wcursor);
-            cursor += PPC_INSTRUCTION_LENGTH;
-            wcursor++;                                  // addi Rb,Rb,lower
-            }
-         else if (displacement != 0)
+         if (displacement < LOWER_IMMED || displacement > UPPER_IMMED)
             {
-            *wcursor = 0x38000000 | (displacement & 0x0000ffff);
-            base->setRegisterFieldRT(wcursor);
-            base->setRegisterFieldRA(wcursor);
-            cursor += PPC_INSTRUCTION_LENGTH;
-            wcursor++;                                  // addi Rb,Rb,disp
+            prevInstruction = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, base, base, cg->hiValue(displacement), prevInstruction);
+            displacement = LO_VALUE(displacement);
             }
 
-         *wcursor = preserve;
-         // just use 0 as RA, base as RB
-         base->setRegisterFieldRB(wcursor);
+         if (displacement != 0)
+            prevInstruction = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, base, base, displacement, prevInstruction);
+
+         self()->setIndexRegister(base);
+         self()->setBaseRegister(cg->machine()->getRealRegister(TR::RealRegister::gr0));
          }
-      else if (displacement<LOWER_IMMED || displacement>UPPER_IMMED ||
-               (self()->offsetRequireWordAlignment() && (displacement & 3 != 0)))
```

Comment:
I think code controlled by this if statement now executes under slightly different condition: there is no check for alignment and index == null.

Reply:
As it turns out, the check for alignment is actually redundant. That case is already fully handled by performing some strange expansion when calling `setOffsetRequiresWordAlignment`, so it's impossible to trigger that condition in practice.

As for `index == NULL`, that's another strange case: it's not valid to have both `index != NULL` and `displacement != 0` unless `isUsingDelayedIndexedForm()` is set, which is already covered by a previous check. To clarify this assumption, I've added an assert in 03d72a33f713b0d22e4f51105ab33e3e21020371 that explicitly checks that this holds.